src/JuliusSweetland.OptiKey.Core/UI/ViewModels/MainViewModel.ServiceEventHandlers.cs

```diff
@@ -2376,44 +2426,175 @@ private void SelectVoice(string voice)
+                    }
+                    else if (commandList.First().StartsWith("KeyUp>"))
+                    {
+                        commandKeyValue = new KeyValue(commandList.First().Substring(6));
+                        Log.InfoFormat("StepListCommand: key up on {0} key", commandKeyValue.String);
+                        await keyboardOutputService.ProcessSingleKeyPress(commandKeyValue.String, KeyPressKeyValue.KeyPressType.Release);
+                        keyStateService.KeyDownStates[commandKeyValue].Value = KeyDownStates.Up;
                     }
-                    else if (vStep.StartsWith("Text>"))
+                    else if (commandList.First().StartsWith("Text>"))
                     {
-                        keyboardOutputService.ProcessSingleKeyText(vStep.Substring(5));
+                        commandKeyValue = new KeyValue(commandList.First().Substring(5));
+                        Log.InfoFormat("StepListCommand: text of '{0}'", commandKeyValue.String);
+                        KeySelectionResult(commandKeyValue, multiKeySelection);
                     }
-                    else if (vStep.StartsWith("Wait>"))
+                    else if (commandList.First().StartsWith("Wait>"))
                     {
-                        System.Threading.Thread.Sleep(int.Parse(vStep.Substring(5)));
+                        var waitMs = int.Parse(commandList.First().Substring(5));
+                        Log.InfoFormat("StepListCommand: wait of {0}ms", waitMs);
+                        await Task.Delay(waitMs);
+                    }
+                }
+
+                if (commandList.Any())
+                {
+                    commandList.RemoveAt(0);
+                }
+                //after returning to the top level at completion of a RunningKey
+                //the key states can be set back to their default states
+                if (nestLevel == 0 && !commandList.Any())
+                {
+                    keyStateService.KeyRunningStates[singleKeyValue].Value = false;
+
+                    if (keyDownList != null && keyDownList.Any())
+                    {
+                        var allKeysUp = true;
+                        foreach (var downKey in keyDownList)
+                        {
+                            var keyUpCandidate = new KeyValue(downKey.Substring(downKey.IndexOf(">")));
+                            if (keyStateService.KeyDownStates[keyUpCandidate].Value != KeyDownStates.Up)
+                            {
+                                allKeysUp = false;
+                                break;
+                            }
+                        }
+                        if (allKeysUp)
+                            keyStateService.KeyDownStates[singleKeyValue].Value = KeyDownStates.Up;
+                    }
+                    else
+                    {
+                        keyStateService.KeyDownStates[singleKeyValue].Value = KeyDownStates.Up;

```

Comment:
This is some more unfinished code that will eventually manage key states when a key is complete.